### DISCUSSION TITLE
After container update, end up with server error ACE00046

### DISCUSSION THREAD

**Discussion by Balubablahblah** [azerothcore/azerothcore-wotlk#10643] [General]
Hello!

Setup is a virtual box running Ubuntu 21.10 with Azerothcore running in containers, all setup through the excellent guide at https://www.azerothcore.org/wiki/install-with-docker

It's been working perfectly up until a week ago where an update left the setup unhappy with its MySQL library versions. Update was done again following the excellent guide above:
```
git pull origin master
./acore.sh docker build
./acore.sh client-data

```
Build goes through, containers start, but no go on connecting the client. Connecting/attaching to the world server show the problem:
```
/azerothcore/apps/startup-scripts/simple-restarter: line 70:   107 Segmentation fault      (core dumped) "./$bin_file"
exit code: 139
worldserver terminated, restarting...
worldserver Terminated after 10 seconds, termination count: : 12
Running without GDB

/azerothcore/src/server/database/Database/DatabaseWorkerPool.cpp:77 in DatabaseWorkerPool FATAL ERROR:
Used MySQL library version (8.0.28 id 80028) does not match the version id used to compile AzerothCore (id 80027).
Search the wiki for ACE00046 in Common Errors (https://www.azerothcore.org/wiki/common-errors).
```

I won't even try to do something in the three containers themself as that is way beyond me. :)

Virtual box have been reverted back to its previous setup, so game server is back up for the huge user base of two persons, but maybe some friendly soul here can help me out on what I need to do to getting a more recent updated version in place?

**Comment by Kitzunu** [comment]
Now I will try to help you but I want to make something clear.

1. I can only tell you the facts
2. I do not know docker

So the facts are that you
A) updated thr MySQL Library version without recompiling your core
B) Updated your MySQL server and recompiled your core but did not update the MySQL Library

**Reply by Balubablahblah** [reply]
Thank you @Kitzunu, but no... I pulled an update on the containers from the repository,  then built new containers that gave me updated MySQL Library and server, and the core was updated in the container.

I'm using the container setup all the way to avoid having to compile anything separately, all to be handled by the container build itself. There is no local MySQL server installed, no local Azerothcore installed, it's all inside containers:

```
CONTAINER ID   IMAGE                                     NAMES
bb6a9a904503   acore/ac-wotlk-worldserver-local:master   azerothcore-wotlk_ac-worldserver_1
e962aa714574   acore/ac-wotlk-authserver-local:master    azerothcore-wotlk_ac-authserver_1
6397150f8a81   mysql:8.0                                 azerothcore-wotlk_ac-database_1

```

**Reply by Kitzunu** [reply]
Like I said, I can only tell you the facts. And that is the only reason as to why the error shows. A missmatch in the mysql client and server

**Comment by pillbug22** [comment]
Having same issue after updating. This is a docker-based install (on top of Ubuntu) that was working well for several months, shut down the server for a few months and came back to it yesterday. Applied OS updates, then applied AC docker updates as per documentation by doing

```
git pull origin master
./acore.sh docker build
```

When then trying to run the docker with normal startup command

`./acore.sh docker start:app`

receiving the same error ACE00046. I understand where on a normal install directly on the host OS this is usually caused by a msql driver mismatch and to rebuild, but it sounds like these are being handled internally by the docker by using the supplied update & rebuild command. If there is a mas-match, then it seems like it should all be handled inside the docker, irrelevant of the host OS, correct?

Are there additional steps to "rebuild" as suggested that are not the provided in the docker update steps?



EDIT:
Reverted back to a snapshot before the updates, and this time only did the docker update steps, did not do any host OS updates just to rule that out. Same result.

```
git pull origin master
./acore.sh docker build
./acore.sh docker start:app
```

Produces the ACE00046 error about MySQL mismatch. Looking a little more closely, it appears this is being thrown by the DB Import container.

```
azerothcore-wotlk-ac-database-1     | 2023-07-13T13:22:59.359023Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Bind-address: '::' port: 33060, socket: /var/run/mysqld/mysqlx.sock
azerothcore-wotlk-ac-database-1     | 2023-07-13T13:22:59.359252Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.31'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server - GPL.
azerothcore-wotlk-ac-db-import-1    | #----------------------------------------------------------------------#
azerothcore-wotlk-ac-db-import-1    |  
azerothcore-wotlk-ac-db-import-1    | >> FATAL ERROR
azerothcore-wotlk-ac-db-import-1    |
azerothcore-wotlk-ac-db-import-1    | # Location '/azerothcore/src/server/database/Database/DatabaseWorkerPool.cpp:78'
azerothcore-wotlk-ac-db-import-1    | # Function 'DatabaseWorkerPool'
azerothcore-wotlk-ac-db-import-1    | # Condition 'isSameClientDB'
azerothcore-wotlk-ac-db-import-1    | # Message 'Used MySQL library version (8.0.33 id 80033) does not match the version id used to compile AzerothCore (id 80031).
azerothcore-wotlk-ac-db-import-1    | Search the wiki for ACE00046 in Common Errors (https://www.azerothcore.org/wiki/common-errors#ace00046).'
azerothcore-wotlk-ac-db-import-1    |  
azerothcore-wotlk-ac-db-import-1    | #----------------------------------------------------------------------#
azerothcore-wotlk-ac-db-import-1 exited with code 139
service "ac-db-import" didn't completed successfully: exit 139
error: Uncaught (in promise) Error: Failed with error: 1, however,
          it's not related to this Deno script directly. An error occurred within
          the script called by the command itself
          throw new Error(`Failed with error: ${status.code}, however,
                ^
    at Command.<anonymous> (file:///home/username/azerothcore-wotlk/apps/docker/docker-cmd.ts:268:17)
    at async Promise.all (index 0)
    at async main (file:///home/username/azerothcore-wotlk/apps/docker/docker-cmd.ts:324:7)
```

EDIT 2: 
This time after reverting back to a working snapshot, updated host OS only. All updates installed, rebooted, ACore docker starts up just fine with no errors (as long as the ACore docker is not updated).

**Reply by Balubablahblah** [reply]
Agreed on that it should be handled by the containers as mysql itself is a separate container, but no. I've been hit with the problem several times, so I ended up setting up a quick routine with doing backup of the databases, scrap the vm, build a fresh one with Ubuntu, fresh Azeroth container setup, restore the databases. The whole process goes quite fast and requires very little attention, with the container build being the most time consuming part.

**Comment by pillbug22** [comment]
@Balubablahblah that sounds like a pretty solid plan. How are you doing the DB exports/imports? I would be concerned that restoring the old DBs on top of an updated fresh setup could undo some of the newer updates to the DB? ...But, I'm assuming that's not an issue or you wouldn't be doing it :-)

**Comment by Balubablahblah** [comment]
Replacing more recent fresh characters and auth databases with old backups have not been a problem so far - any changes for the more recent version are automatically applied on startup.

Both old and new VM got access to a shared folder on the host system.

**Old VM** - make sure everything works, then stop everything but the mysql container, backup:
```
sudo ./acore.sh docker start:app:d
sudo docker container stop azerothcore-wotlk-ac-authserver-1
sudo docker container stop azerothcore-wotlk-ac-worldserver-1
sudo docker exec azerothcore-wotlk-ac-database-1 mysqldump -u root --password=**** acore_characters > /mnt/vmshare/acore_characters.sql
sudo docker exec azerothcore-wotlk-ac-database-1 mysqldump -u root --password=**** acore_auth > /mnt/vmshare/acore_auth.sql
```

**New VM** - make sure everything works, then stop everything but the mysql container, restore:
```
sudo ./acore.sh docker start:app:d
sudo docker container stop azerothcore-wotlk-ac-authserver-1
sudo docker container stop azerothcore-wotlk-ac-worldserver-1
sudo cat /mnt/vmshare/acore_characters.sql | sudo docker exec -i azerothcore-wotlk-ac-database-1 mysql -u root --password=**** acore_characters
sudo cat /mnt/vmshare/acore_auth.sql | sudo docker exec -i azerothcore-wotlk-ac-database-1 mysql -u root --password=**** acore_auth
```

**Comment by pillbug22** [comment]
Came back to try this out yesterday and noticed the ACore docker instructions page is updated. That said, your steps worked great on a new install - brought in my old DBs without any issues. Will see if the updated docker instructions (including a new GitHub path) allow an in-place update now. Definitely have this as a backup plan, though. Thanks for sharing!

**Reply by Balubablahblah** [reply]
Happy to help!
